File: backend/app/pipeline/preprocess.py
import tempfile
from pathlib import Path

import cv2
import numpy as np


def preprocess_floor_plan(input_path: str) -> str:
    """
    Preprocess a floor plan image for VLM parsing.
    Returns the path to the processed image.
    """
    img = cv2.imread(input_path)
    if img is None:
        raise ValueError(f"Could not read image: {input_path}")

    # Convert to grayscale for processing
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # Auto-rotate correction using Hough line detection
    img = _correct_rotation(img, gray)

    # Crop to content area
    img = _crop_content(img)

    # Resize if too large
    img = _resize_if_large(img, max_size=2048)

    # Save processed image
    output_path = _save_processed(input_path, img)
    return output_path


def _correct_rotation(img: np.ndarray, gray: np.ndarray) -> np.ndarray:
    """Detect dominant lines and correct rotation angle."""
    try:
        edges = cv2.Canny(gray, 50, 150, apertureSize=3)
        lines = cv2.HoughLinesP(edges, 1, np.pi / 180, threshold=100, minLineLength=100, maxLineGap=10)

        if lines is None or len(lines) == 0:
            return img

        angles = []
        for line in lines:
            x1, y1, x2, y2 = line[0]
            angle = np.degrees(np.arctan2(y2 - y1, x2 - x1))
            # Only consider near-horizontal/vertical lines
            if abs(angle) < 45 or abs(angle - 90) < 45 or abs(angle + 90) < 45:
                angles.append(angle)

        if not angles:
            return img

        median_angle = np.median(angles)

        # Only correct small rotations
        if abs(median_angle) > 5 and abs(median_angle) < 85:
            return img

        if abs(median_angle) < 2:
            return img

        h, w = img.shape[:2]
        center = (w // 2, h // 2)
        M = cv2.getRotationMatrix2D(center, median_angle, 1.0)
        img = cv2.warpAffine(img, M, (w, h), flags=cv2.INTER_CUBIC, borderMode=cv2.BORDER_REPLICATE)
    except Exception:
        pass

    return img


def _crop_content(img: np.ndarray) -> np.ndarray:
    """Crop to the content area, removing large white borders."""
    try:
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        # Threshold to find non-white areas
        _, thresh = cv2.threshold(gray, 240, 255, cv2.THRESH_BINARY_INV)
        # Dilate to connect nearby elements
        kernel = np.ones((5, 5), np.uint8)
        dilated = cv2.dilate(thresh, kernel, iterations=2)
        # Find bounding box
        coords = cv2.findNonZero(dilated)
        if coords is not None:
            x, y, w, h = cv2.boundingRect(coords)
            # Add padding
            pad = 20
            x = max(0, x - pad)
            y = max(0, y - pad)
            w = min(img.shape[1] - x, w + 2 * pad)
            h = min(img.shape[0] - y, h + 2 * pad)
            if w > 100 and h > 100:
                img = img[y : y + h, x : x + w]
    except Exception:
        pass

    return img


def _resize_if_large(img: np.ndarray, max_size: int = 2048) -> np.ndarray:
    """Resize image if its longest side exceeds max_size."""
    h, w = img.shape[:2]
    longest = max(h, w)
    if longest <= max_size:
        return img

    scale = max_size / longest
    new_w = int(w * scale)
    new_h = int(h * scale)
    return cv2.resize(img, (new_w, new_h), interpolation=cv2.INTER_AREA)


def _save_processed(original_path: str, img: np.ndarray) -> str:
    """Save the processed image to a temporary file."""
    src = Path(original_path)
    suffix = src.suffix or ".png"
    fd, output_path = tempfile.mkstemp(suffix=suffix, prefix="planova_processed_")
    import os
    os.close(fd)
    cv2.imwrite(output_path, img)
    return output_path

File: src-tauri/src/pipeline/preprocess.rs
use std::path::Path;

use image::DynamicImage;

pub fn preprocess_floor_plan(input_path: &str) -> Result<String, String> {
    let img = image::open(input_path).map_err(|e| format!("Could not read image: {e}"))?;
    log::info!("Input image: {}x{}", img.width(), img.height());

    // Resize first to cap memory usage before heavy processing
    let img = resize_if_large(img, 2048);

    // Convert to grayscale for processing
    let gray = img.to_luma8();

    // Auto-rotate correction
    let img = correct_rotation(img, &gray);

    // Crop to content area
    let img = crop_content(img);

    // Save to temp file
    let ext = Path::new(input_path)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("png");
    let output_path = std::env::temp_dir().join(format!(
        "planova_processed_{}.{}",
        uuid::Uuid::new_v4().simple(),
        ext
    ));
    img.save(&output_path)
        .map_err(|e| format!("Failed to save processed image: {e}"))?;

    Ok(output_path.to_string_lossy().to_string())
}

fn correct_rotation(img: DynamicImage, gray: &image::GrayImage) -> DynamicImage {
    use imageproc::edges::canny;

    // Canny edge detection
    let edges = canny(gray, 50.0, 150.0);

    // Simple Hough-like line detection: collect angles from edge pixels
    // using gradient direction sampling
    let mut angles: Vec<f64> = Vec::new();
    let (w, h) = edges.dimensions();

    // Sample edge pixels and compute local gradient angles
    for y in 1..h - 1 {
        for x in 1..w - 1 {
            if edges.get_pixel(x, y)[0] > 0 {
                // Compute gradient using Sobel-like approximation
                let gx = gray.get_pixel(x + 1, y)[0] as f64 - gray.get_pixel(x - 1, y)[0] as f64;
                let gy = gray.get_pixel(x, y + 1)[0] as f64 - gray.get_pixel(x, y - 1)[0] as f64;

                if gx.abs() + gy.abs() > 10.0 {
                    let angle = gy.atan2(gx).to_degrees();
                    // Normalize to near-horizontal/vertical
                    let norm_angle = if angle > 45.0 {
                        angle - 90.0
                    } else if angle < -45.0 {
                        angle + 90.0
                    } else {
                        angle
                    };
                    if norm_angle.abs() < 5.0 {
                        angles.push(norm_angle);
                    }
                }
            }
        }
    }

    if angles.is_empty() {
        return img;
    }

    // Compute median angle
    angles.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let median_angle = angles[angles.len() / 2];

    if median_angle.abs() < 0.5 {
        return img;
    }

    log::info!("Rotation correction: {:.2} degrees", median_angle);

    use imageproc::geometric_transformations::{rotate_about_center, Interpolation};

    let rgba = img.to_rgba8();
    let rad = (median_angle as f32).to_radians();
    let rotated = rotate_about_center(
        &rgba,
        rad,
        Interpolation::Bilinear,
        image::Rgba([255, 255, 255, 255]),
    );

    DynamicImage::ImageRgba8(rotated)
}

fn crop_content(img: DynamicImage) -> DynamicImage {
    let gray = img.to_luma8();
    let (w, h) = gray.dimensions();

    // Find non-white pixels (threshold at 240)
    let mut min_x = w;
    let mut min_y = h;
    let mut max_x = 0u32;
    let mut max_y = 0u32;
    let mut found = false;

    for y in 0..h {
        for x in 0..w {
            if gray.get_pixel(x, y)[0] < 240 {
                min_x = min_x.min(x);
                min_y = min_y.min(y);
                max_x = max_x.max(x);
                max_y = max_y.max(y);
                found = true;
            }
        }
    }

    if !found || max_x <= min_x || max_y <= min_y {
        return img;
    }

    // Add padding (20px)
    let pad = 20u32;
    let x0 = min_x.saturating_sub(pad);
    let y0 = min_y.saturating_sub(pad);
    let x1 = (max_x + pad).min(w.saturating_sub(1));
    let y1 = (max_y + pad).min(h.saturating_sub(1));

    if x1 <= x0 || y1 <= y0 {
        return img;
    }
    let cw = x1 - x0;
    let ch = y1 - y0;

    if cw < 100 || ch < 100 {
        return img;
    }

    log::info!("Cropped: {}x{} -> {}x{}", w, h, cw, ch);
    img.crop_imm(x0, y0, cw, ch)
}

fn resize_if_large(img: DynamicImage, max_size: u32) -> DynamicImage {
    let (w, h) = (img.width(), img.height());
    let longest = w.max(h);
    if longest <= max_size {
        return img;
    }

    let scale = max_size as f64 / longest as f64;
    let new_w = (w as f64 * scale) as u32;
    let new_h = (h as f64 * scale) as u32;

    log::info!("Resized: {}x{} -> {}x{}", w, h, new_w, new_h);
    img.resize_exact(new_w, new_h, image::imageops::FilterType::Lanczos3)
}
